### DISCUSSION TITLE
magick: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG

### DISCUSSION THREAD

**Discussion by finefoot** [ImageMagick/ImageMagick#6292] [Help] [answered]
### ImageMagick version

7.1.1-8

### Operating system, version and so on

Debian 11

### Description

I have a question related to [#2730](https://github.com/ImageMagick/ImageMagick/discussions/2730).

Here is the example PDF file that I'm using: [example.pdf](https://github.com/ImageMagick/ImageMagick/files/11378072/example.pdf)

When I want to convert this PDF to PNG, I get the following warning:

    $ magick example.pdf output.png
    magick: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `output.png' @ warning/png.c/MagickPNGWarningHandler/1526.
    $ 

As far as I understand, ImageMagick includes the original PDF (RGB color) ICC profile into the (grayscale) PNG output file which produces this (harmless) warning.

If I strip profiles, the warning disappears:

    $ magick example.pdf -strip output.png
    $ 

However, a few discussion threads suggest changing the colorspace, to get rid of the warning:
- https://superuser.com/questions/1291967/error-in-converting-a-pdf-to-a-png-in-imagemagick
- https://unix.stackexchange.com/questions/485292/rgb-color-space-not-permitted-on-grayscale

So if I try that out and convert to RGB colorspace, the warning indeed disappears:

    $ magick example.pdf -colorspace RGB output.png
    $ 

However, for sRGB colorspace, the warning returns:

    $ magick example.pdf -colorspace sRGB output.png
    magick: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `output.png' @ warning/png.c/MagickPNGWarningHandler/1526.
    $ 

If I take a detour via RGB and then back to sRGB colorspace, no warning again:

    $ magick example.pdf -colorspace RGB -colorspace sRGB output.png
    $ 

I want to understand why the warning sometimes shows up and sometimes doesn't. Why does changing the colorspace to RGB help here?

```
$ magick -version
Version: ImageMagick 7.1.1-8 Q16-HDRI x86_64 d92cb0e65:20230421 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5) 
Delegates (built-in): bzlib djvu fontconfig freetype heic jbig jng jpeg lcms lqr lzma openexr png raqm tiff webp x xml zlib
Compiler: gcc (7.5)
```

**Comment by fmw42** [comment]
Your PDF is grayscale. You are trying to put a color profile on a grayscale PNG and PNG does not like that. It is only a warning not an error.

**Comment by snibgo** [comment]
The warning originates from libpng. IM tells libpng to convert the image to PNG format. If libpng recognises that the ICC profile is for three channels (R,G and B) but the image is grayscale, it raises the warning. But if the image has three channels, libpng may not realise the channels contain identical values.

I don't know the internals of libpng, and don't know when (or if) it checks for channel equality.

**Reply by finefoot** [reply]
Thank you very much for your reply. :)

I'm sorry, but I still don't understand. Maybe this is a stupid question, but how does that explain the different behavior regarding RGB vs. sRGB? Or are you saying that this question is off-topic regarding ImageMagick and I must ask elsewhere?

**Comment by snibgo** [comment]
In this command ...
```
magick example.pdf -verbose info:
```
... notice that the text output includes:
```
  Profiles:
    Profile-icc: 2576 bytes
```

If the image that is input to the operation "`-colorspace sRGB`" is already sRGB and has three channels (R,G and B) then the operation does nothing, so has no effect on what libpng does.

By contrast, the operation "`-colorspace RGB`" changes pixel values, and also removes the sRGB ICC profile. You can see this with:
```
magick example.pdf -colorspace RGB -verbose info:
```
The text output no longer contains:
```
  Profiles:
    Profile-icc: 2576 bytes
```
(Removing the profile is good, because it would claim the image was non-linear sRGB instead of linear RGB.)

So the command ...
```
magick example.pdf -colorspace RGB -colorspace sRGB output.png
```
... results in no ICC profile being sent to libpng, so it has nothing to warn about.

**Reply by finefoot** [reply]
That is very helpful! Thank you. :)

So, the PDF file is in colorspace sRGB, initially, right? So `-colorspace RGB` does a conversion to RGB. But why does that action also remove the profile? In other words: If I run `magick input -colorspace RGB -colorspace sRGB output` on an input file with sRGB colorspace, it is unexpected to me, that input file and output file should differ.

I did read the description of the `-colorspace` option at https://imagemagick.org/script/command-line-options.php#colorspace but it didn't explain the observed result of stripping the profile. Where is my misunderstanding? How can I explain what's happening?

**Comment by snibgo** [comment, marked as answer]
> So, the PDF file is in colorspace sRGB, initially, right?

As far as ImageMagick is concerned, yes, the image is in sRGB colorspace. But to be more precise: Ghostscript makes an image that contains an embedded ICC color profile that claims to be sRGB. 

However, not all sRGB ICC profiles are the same. IM is distributed with `sRGB.icc`. Your image has all pixels (RGB)=(1,1,1) on a scale of 0 to 1. If we convert your image to that profile:
```
magick example.pdf -profile sRGB.icc -verbose info:

  Channel statistics:
    Pixels: 493240
    Red:
      min: 65621.9  (1.00133)
      max: 65621.9 (1.00133)
    Green:
      min: 65404.7  (0.998011)
      max: 65404.7 (0.998011)
    Blue:
      min: 65492.5  (0.999352)
      max: 65492.5 (0.999352)
```
Oops. This is why, when an image contains an embedded profile, for best precision we shouldn't blindly strip the profile. For example, if we want a linear version of sRGB, we should convert it using `-profile` not `-colorspace`.

> If I run magick input -colorspace RGB -colorspace sRGB output on an input file with sRGB colorspace, it is unexpected to me, that input file and output file should differ.

If there was no embedded profile, they would be the same.

> I did read the description of the -colorspace option at https://imagemagick.org/script/command-line-options.php#colorspace but it didn't explain the observed result of stripping the profile. Where is my misunderstanding? How can I explain what's happening?

The documentation doesn't explain all the actions of the code. (If it did, it would be almost as large as the code.)

In the source code `MagickCore/colorspace.c`, function `TransformImageColorspace()`, we find:
```
  (void) DeleteImageProfile(image,"icc");
  (void) DeleteImageProfile(image,"icm");
```
When that function is called, and changes the colorspace, all ICC and ICM profiles are stripped.

**Comment by finefoot** [comment]
That's some great explanations. Thank you very much, I'm learning a lot! :)

So now I understand that changing the colorspace removes the profile and therefore the warning disappears. Let's say that changing the colorspace isn't really what I want to do. I only want to remove the profile. I found `-strip` which removes profiles, but unfortunately also removes other metadata, too. Then I found `-define profile:skip=`, which is also mentioned in one of the answers linked above. Unfortunately, it doesn't do what I expected:

```
$ magick example.pdf -define profile:skip=icc -verbose info: | grep -A1 "Profiles:"
  Profiles:
    Profile-icc: 2576 bytes
$ 
```

Why does this not skip the profile? I also tried it with `skip="*"`. And I still get the warning when converting to PNG.

However, I also found `+profile`, which (apparently) does what I want:

```
$ magick example.pdf +profile icc -verbose info: | grep -A1 "Profiles:"
$ 
```

So does `+profile icc` remove _only_ the ICC profile and nothing else? And what's the difference to `-define profile:skip=icc`?

**Comment by snibgo** [comment]
Most IM options are either *operations* or *settings*.

A operation changes images that are in memory. Most operations change all the images that are in the current list. Examples: "`-strip`" and "`+profile icc`".

A setting doesn't change any images. Instead, it affects the action taken by following operations. (A "following" operation occurs after the setting within the command.) Defines (see https://imagemagick.org/script/defines.php) are settings. "`-define profile:skip=icc`" alters the actions of following operations that read images. For example, reading with no `-define`:
```
magick example.pdf -verbose info: |grep -i profile

  Profiles:
    Profile-icc: 2576 bytes
    icc:description: Artifex Software sRGB ICC Profile
```
Or we can skip all ICC profiles:
```
magick -define profile:skip=icc example.pdf -verbose info: |grep -i profile
```
Note that the `-define` must occur *before* the PDF is read.

> So does `+profile icc` remove *only* the ICC profile and nothing else?

Yes.

**Reply by finefoot** [reply]
> Note that the -define must occur before the PDF is read.

🤦‍♂️ Ugh, of  course. My bad, sorry.

Thank you for the support and your excellent explanations.
